app/main.py

```diff
@@ -86,128 +86,51 @@
 }
 
 
-def battle(knightsConfig):
-    # BATTLE PREPARATIONS:
+def apply_effects(knight: dict) -> None:
+    """Apply all effects (armor, weapon, potion) to a knight."""
+    # Initialize protection
+    knight["protection"] = 0
 
-    # lancelot
-    lancelot = knightsConfig["lancelot"]
+    # Apply armor
+    for armor_piece in knight["armour"]:
+        knight["protection"] += armor_piece["protection"]
 
-    # apply armour
-    lancelot["protection"] = 0
-    for a in lancelot["armour"]:
-        lancelot["protection"] += a["protection"]
+    # Apply weapon
+    knight["power"] += knight["weapon"]["power"]
 
-    # apply weapon
-    lancelot["power"] += lancelot["weapon"]["power"]
+    # Apply potion if exists
+    if knight["potion"] is not None:
+        for stat, value in knight["potion"]["effect"].items():
+            knight[stat] += value
 
-    # apply potion if exist
-    if lancelot["potion"] is not None:
-        if "power" in lancelot["potion"]["effect"]:
-            lancelot["power"] += lancelot["potion"]["effect"]["power"]
 
-        if "protection" in lancelot["potion"]["effect"]:
-            lancelot["protection"] += lancelot["potion"]["effect"]["protection"]
+def process_battle(knight1: dict, knight2: dict) -> None:
+    """Process a battle between two knights."""
+    # Calculate damage
+    knight1["hp"] -= max(0, knight2["power"] - knight1["protection"])
+    knight2["hp"] -= max(0, knight1["power"] - knight2["protection"])
 
-        if "hp" in lancelot["potion"]["effect"]:
-            lancelot["hp"] += lancelot["potion"]["effect"]["hp"]
+    # Ensure HP doesn't go below 0
+    knight1["hp"] = max(0, knight1["hp"])
+    knight2["hp"] = max(0, knight2["hp"])
 
-    # arthur
-    arthur = knightsConfig["arthur"]
 
-    # apply armour
-    arthur["protection"] = 0
-    for a in arthur["armour"]:
-        arthur["protection"] += a["protection"]
+def battle(knightsconfig: dict) -> dict:
+    # Create working copies of the knights
+    knights = {name: knight.copy() for name, knight in knightsconfig.items()}
```

Comment:
The 'copy' method creates a shallow copy of the dictionary. Since the knight's attributes include nested dictionaries (like 'weapon' and 'potion'), a deep copy should be used to avoid modifying the original 'KNIGHTS' data. Consider using 'copy.deepcopy()' from the 'copy' module.